### ISSUE TITLE
chromaprint fpcalc needs to be added to PATH

### ISSUE THREAD

**Issue by adispi** [SynoCommunity/spksrc#3687]
please see this thread:
https://github.com/lidarr/Lidarr/issues/746

**Comment by ta264** [comment]
For what it's worth, it seems to be trying to add a link:
https://github.com/SynoCommunity/spksrc/blob/master/spk/chromaprint/src/service-setup.sh#L4

But given the reports in https://github.com/lidarr/Lidarr/issues/746 and procedure for fixing it can't be working correctly

**Comment by BenjV** [comment]
Adding binaries to the PATH is against the DSM architecture and will be gone after every update of DSM.
If Lidarr wants/needs to use an other packages it should provide functionality to add the location of that  package so it can be found.

Adding a link is not the same as adding something to the PATH environment variable and the link you found is and incorrect link to a location where DSM stores some binaries. 
That link should not be created in the package and does not serve any purpose.

**Comment by ta264** [comment]
So you're saying that if you install a package in DSM then you can only use the installed binaries if you know the full path? They don't get linked into a directory already on the users PATH variable? That seems very strange coming from a Linux background. Obviously I agree it shouldn't be adding the absolute path to the fpcalc binary to the path environment variable.

**Comment by ta264** [comment]
Is the issue that mkdir is missing for chromaprint so if the directory doesn't already exist then `ln` fails? For example from beets
https://github.com/SynoCommunity/spksrc/blob/master/spk/beets/src/service-setup.sh#L14

**Comment by ymartin59** [comment]
@ta264 Thank you for diagnostic... I added this probably too fast.. Problem is that chromaprint is heavy to rebuild for such a small patch.

I propose following work-around: run `sudo mkdir -p /usr/local/bin` from ssh as any administrator account before installing chromaprint package...

**Comment by BenjV** [comment]
No that is completely wrong,
The folder /usr/local/bin should not be used for packages that is a DSM folder and most likely all will be gone if DSM get updated.

If an application like Lidarr needs a external application it should provide a way to find it and not by messing around in system folders.

So either configuration file or in an own environment variable.
And due to the fact that all packages on DSM are running as an hidden users without a user environment it is not possible to use user PATH variable.

**Comment by ta264** [comment]
In that case fpcalc should be added to the lidarr specific path as sort of the lidarr package just like mono is
https://github.com/SynoCommunity/spksrc/blob/master/spk/lidarr/src/service-setup.sh#L19

**Comment by hgy59** [comment]
@BenjV there is an official documented way to create links for spk resources into subfolders of /usr/local like bin, lib, etc.
I found it in the synology package developer guide in section **4.10.1 Resource Specification**.

I don't know why `synocommunity/spksrc` does not use this, since it only needs a file named `resource` in the `conf` folder of the package.
The resource file is in JSON and may contain several different `workers` for port configuration, php.ini creation, service creation and more.
The worker to create links in /usr/local is called **/usr/local linker** (section 4.10.4.1)

The documentation shows a sample resource file with a linker worker for a perl package:
```JSON
"usr-local-linker": {
  "bin": ["usr/bin/a2p", "usr/bin/perl"],
  "lib": ["lib/perl5"]
}
```
The path of a target to link is relative to the package target folder.


@ymartin59  maybe it is time to enhance the process for spk generation with resource configuration.

A good overview of the resource workers is given in section **4.10.2 Timing**
It explains that the links are created when a package changes from disabled to enabled and the links are removed when the packages changes from enabled to disabled. The most important benefit is, that this is handled by the DSM framework.

Now I propose to update the `lidarr` and `mono` packages with a `conf/resource` file to create the required links.
And for `lidarr` the path modification should be removed from `service-setup.sh`.

If this works like expected, there is a lot of work to redesign all the packages that currently create links a different way or modify the PATH. 
- found 78 packages that create links in service-setup.sh or installer.sh
- found 67 packages that modify the PATH in service-setup.sh oder installer.sh
- some packages have a script dsm-control.sh where the PATH is modified too

Let's redesign the spk creation to get rid of all the problems with missing paths after DSM updates!

**Comment by BenjV** [comment]
Nice catch, but in my honest opinion not to be used.
That functionality is not only not used by the SynoCommunity but also not in packages from Synology themselves.
This is just another sloppy way of creating a search path to find binaries en very much outdated architecture.
These search paths are only designed for users of the commandline, so they don't have to deal with typing a long path when they use system commands and should not be use for applications.
If an application needs another application or binary, then such application should provide a way to add the location where to find it and not only using a search path.

The fact that a lot of package are changing the PATH is most likely due to package builders that don't understand the DSM architecture and just did wat they thought was correct and most likely not needed at all because there are very little application that need it.
If they need another package then it should also be added to the INFO file to create an install dependency, so if that is not the case adding it to the PATH environment variable is useless.

The fact that lidarr is a bad designed application that does not have a way to add the location of needed binaries but rely on the PATH is no reason to go chancing lots of packages.
Changing lots of packages means lots of testing and that can only be done by people who use those applications.

If it is needed for lidarr to function I propose to only add that binary to the bin folder of lidarr package so it can be found by lidarr without a search path.

**Comment by ymartin59** [comment]
@hgy59 I am aware of resource configuration described in SDK documentation. My idea is to introduce its support at first for database usage - user/schema creation for applications - with choice for mysql or postgresql from wizard - typically for ownCloud.

**Comment by hgy59** [comment]
@BenjV resource file of the synology mono package (Mono-x86_64-4.6.2-0096.spk)
```JSON
{
	"usr-local-linker": {
		"bin": ["usr/local/bin/mono", "usr/local/bin/mcs", "usr/local/bin/monodis", "usr/local/bin/mono-sgen", "usr/local/bin/pedump", "usr/local/bin/xsp4", "usr/local/bin/gacutil", "usr/local/bin/gacutil2", "usr/local/bin/mod-mono-server4"],
		"lib": ["usr/local/lib/libMonoPosixHelper.a", "usr/local/lib/libMonoPosixHelper.la", "usr/local/lib/libMonoPosixHelper.so", "usr/local/lib/mono"]
	}
}
```

**Comment by hgy59** [comment]
This issue is solved with #3852.
